probe/core/src/main/java/org/jboss/weld/probe/ProbeServlet.java

```diff
@@ -0,0 +1,375 @@
+import java.io.UnsupportedEncodingException;
+
+import javax.enterprise.inject.spi.Bean;
+import javax.enterprise.inject.spi.ObserverMethod;
+import javax.inject.Inject;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.jboss.weld.manager.BeanManagerImpl;
+import org.jboss.weld.probe.Queries.BeanFilters;
+import org.jboss.weld.probe.Queries.Filters;
+import org.jboss.weld.probe.Queries.InvocationsFilters;
+import org.jboss.weld.probe.Queries.ObserverFilters;
+
+/**
+ * A simple Probe REST API implementation.
+ *
+ * <p>
+ * An integrator is required to register this servlet if appropriate.
+ * </p>
+ *
+ * @author Martin Kouba
+ */
+public class ProbeServlet extends HttpServlet {
+
+    public static final String DEFAULT_URL_PATTERN = "/weld-probe/*";
+
+    private static final String PARAM_TRANSIENT_DEPENDENCIES = "transientDependencies";
+
+    private static final String PARAM_TRANSIENT_DEPENDENTS = "transientDependents";
+
+    private static final String PARAM_PAGE = "page";
+
+    private static final String PARAM_FILTERS = "filters";
+
+    private static final String FILE_CLIENT_HTML = "probe.html";
+
+    private static final int DEFAULT_BUFFER = 1024 * 8;
+
+    private static final long serialVersionUID = -881168492978480911L;
+
+    @Inject
+    private BeanManagerImpl beanManager;
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+
+        Probe probe = getProbe();
+        ResourcePath resourcePath;
+        String[] pathInfoParts = getPathInfoParts(req);
+
+        if (pathInfoParts == null) {
+            resourcePath = ResourcePath.ROOT;
+        } else {
+            resourcePath = match(pathInfoParts);
+            if (resourcePath == null) {
+                resp.sendError(HttpServletResponse.SC_NOT_FOUND);
+                return;
+            }
+        }
+
+        ProbeLogger.LOG.resourcePathMatched(resourcePath, req.getPathInfo());
+
+        if (ROOT.equals(resourcePath) || CLIENT_RESOURCE.equals(resourcePath)) {
+
+            // Root resource - a default HTML client
+            String resourceName;
+            if (pathInfoParts == null) {
+                resourceName = FILE_CLIENT_HTML;
+            } else {
+                resourceName = ROOT.equals(resourcePath) ? pathInfoParts[0] : pathInfoParts[1];
+            }
+
+            String contentType = detectContentType(resourceName);
+
+            if (IMG_PNG.equals(contentType)) {
+                if (!writeResource(resourceName, resp.getOutputStream())) {
+                    resp.sendError(HttpServletResponse.SC_NOT_FOUND);
+                }
+            } else {
+                // All other content types are text-based
+                String content = getResourceAsString(resourceName);
+                if (content == null) {
```

Comment:
What do you mean? `getResourceAsString()` returns null if there is no resource found.